### PR TITLE
New script generating release notes

### LINKED ISSUE #5001: Make `generate_release_notes.py` much faster
Right now `generate_release_notes.py` takes half an hour or so to query the 192 relevant PRs from the GitHub website.

That's really bad, and I am sure we can do better. Indeed using the [`gh` command line tool](https://cli.github.com) I can easily execute the relevant query in about 5 seconds, producing JSON that's not so far from what we need. Note that unlike our existing script, I am using the `merged` filter / `mergedAt` property (instead of `closed` / `closedAt`) which reduces the number of matches server side, and then I also tell GitHub to filter out anything with a certain label. This alone is not explaining the several orders of magnitude difference in performance, but they do contribute.
```
gh pr list --search 'merged:>=2019-09-09 -label:"release notes: not needed"' --json number,title,closedAt,labels,mergedAt --limit 200
```

(Actually, I first run this with a smaller limit to find out how many matches there are, then set the limit high enough to get all of them).

### PR DESCRIPTION
**Pull request by fingolfin**
Resolves #5001

### CODE REVIEW

File: dev/releases/create_stable_branch.py [hunk 1]
```diff
@@ -14,81 +14,107 @@
 # TODO: implement parts of the steps described in
 # <https://github.com/gap-system/gap-distribution/blob/master/DistributionUpdate/STABLE_BRANCH_CHECKLIST.md>
 
-from utils import *
+from utils import error, notice, warning, patchfile
```
Comment (james-d-mitchell) [inline dev/releases/create_stable_branch.py:17]:
```suggestion
from utils import error, notice, patchfile
```
`warning` isn't used

File: dev/releases/create_stable_branch.py [hunk 2]
```diff
@@ -14,81 +14,107 @@
...
+    stable_branch = "stable-" + gapversion
 
-# TODO: error out if the branch already exists
+    notice("Checking prerequisites")
+    utils.verify_command_available("git")
+    utils.verify_git_repo()
+    utils.verify_git_clean()
 
-# TODO: Create a pair of labels for GitHub issues called backport-to-X.Y and backport-to-X.Y-DONE.
+    notice(f"Switching to master branch")
```
Comment (james-d-mitchell) [inline dev/releases/create_stable_branch.py:44]:
```suggestion
    notice("Switching to master branch")
```
Nothing is formatted, so the "f" is superfluous.

File: dev/releases/create_stable_branch.py [hunk 3]
```diff
@@ -14,81 +14,107 @@
...
+    subprocess.run(["git", "switch", "master"], check=True)
 
-# TODO: Create a GitHub milestone for GAP X.Y.0 release.
+    notice(f"Ensure branch is up-to-date")
```
Comment (james-d-mitchell) [inline dev/releases/create_stable_branch.py:47]:
```suggestion
    notice("Ensure branch is up-to-date")
```
Same again

File: dev/releases/create_stable_branch.py [hunk 4]
```diff
@@ -14,81 +14,107 @@
...
-# list of files which (potentially) are updated
-files = [
+    # list of files which (potentially) are updated
+    files = [
         "CITATION",
         "configure.ac",
         "doc/versiondata",
+    ]
+
+    notice("Updating version to {nextgapversion} on master branch")
```
Comment (james-d-mitchell) [inline dev/releases/create_stable_branch.py:61]:
```suggestion
    notice(f"Updating version to {nextgapversion} on master branch")
```

Missing `f`

**Comment by james-d-mitchell** [comment]
There are a bunch of (minor) issues with the python code that could be improved (unused imports, non-pythonic code like `x == None` rather than `x is None` etc), but the scripts I ran locally seemed to work ok (qualifier: I'm not completely sure what you are trying to achieve with the files in this PR, and haven't examined everything in detail). I'd be happy to resolve these issues, if that's helpful? I commented on a few issues but this list isn't exhaustive.

Some questions:

1. is there anything in the CI that tests these scripts?
2. if not, then how would one go about testing that they work as intended (other than just running them and attempting to make a release)?

**Comment by fingolfin** [comment]
> There are a bunch of (minor) issues with the python code that could be improved (unused imports, non-pythonic code like `x == None` rather than `x is None` etc), but the scripts I ran locally seemed to work ok (qualifier: I'm not completely sure what you are trying to achieve with the files in this PR, and haven't examined everything in detail). I'd be happy to resolve these issues, if that's helpful? I commented on a few issues but this list isn't exhaustive.

Sure, go ahead.

> 
> Some questions:
> 
> 1. is there anything in the CI that tests these scripts?

Not for the scripts generating release notes and creating the stable branch.

But of course the scripts that actually make the release are tested, which is good because I touched them a little bit, too, and testing them manually is quite annoying.

> 2. if not, then how would one go about testing that they work as intended (other than just running them and attempting to make a release)?

So far I just didn't think it's worth the effort to try to automate testing for these. But if someone else were to spend efforts on it, of course I wouldn't mind... In each case the only somewhat OK way to test them I can think of is to use some fixed pre-defined set of inputs for them, and store a reference output in the repository, and compare against that.

- for `generate_package_release_notes.py` one could e.g. run `generate_package_release_notes.py 4.12.1 4.12.2` and verify the output doesn't change
- for `release_notes.py` somewhat similar, but I can't see how to do it w/o adding something to the script: specifically, I think one would have to add a command line argument that specifies an upper bound for when the issues were created (or possibly an upper bound on the issue number, or whatever); we wouldn't use that bound in actual usage, but for tests it would allow running, say `./release_notes 4.12.0 --before=SOMEDATE` and `./release_notes 4.12.1 --before=SOMEDATE` with hopefully stable output (one should test both a "minor" and a "major" release as they have slightly different queries)
- for `create_stable_branch.py` I guess one would have to clone `gap` with full history (including at least `master` and `stable-4.12`; then `git reset --hard master SOME_FIXED_COMMIT`, say just before stable-4.12 was branched off; then add a way to disable the `git pull --ff-only` in the script; then ensure the `4.13dev` tag and `stable-4.12` branch are not there (just `git branch -D stable-4.12`) and run `./create_stable_branch.py 13` -- then verify that `git diff master SHA-OF-PUBLIC-v4.13dev-TAG` and `git diff stable-4.12 SHA-OF-PUBLIC-v4.12.0-TAG` report no diffs. And have sensible commit messages.


But again, I don't think it's that important to have CI tests for them: I tested them extensively while preparing this PR. Otherwise they are usually never changed. Of course they are also rarely used (just a few times per year for releases) and if they break it's holding up a release, but the risk is low overall, typically it would be some URL in the package distro changing (not that they should change, but that's one of the few external things that could change w/o anyone touching the scripts).

**Comment by fingolfin** [comment]
BTW ideally also `generate_release_notes.py` would be merged into `release_notes.py` to resolve #5004. The main "blockers" there are:
- `generate_release_notes.py` currently needs to be given the previous GAP release. Of course if we try to release 4.12.3 it is trivial to compute that 4.12.2 was the previous release. But for 4.13.0 it requires a little more work -- e.g. we may assume that all public tags have been fetched check for the presence of a tag `v4.12.0`, `v4.12.1`, `v4.12.2`, ... until it find one that does not exist (here `v4.12.3`) and then assumes that `v4.12.2` is the last release.
- for testing it would be nice if I could give it a version that's not yet been tagged in the package distro (so "4.13.0" right now) and it would automatically detect this and instead of trying to fetch JSON files for `v4.13.0`, instead would fetch them for `latest`. Or maybe instead of "detecting" this (how exactly would it do that anyway?) one could just add an optional command line argument to force this (perhaps just as simple as `./release_notes.py 4.13.0 latest`?)

**Comment by fingolfin** [comment]
@james-d-mitchell how about we merge this PR, and you add your changes in a fresh PR to this repo? That might be easier to manage and increases visibility for others :-)

**Comment by james-d-mitchell** [comment]
Sure thing @fingolfin sounds good

**Review by james-d-mitchell — APPROVED** [review]
LGTM

**Pull request merged**